### PR TITLE
Keep built Android APKs as artifacts

### PR DESCRIPTION
**Pull request by kaidokert**
Preserves built APKs in a single bundle for all architectures.
This is intended to keep APKs with every release.
    
b/290121545
b/318577880

### CODE REVIEW

File: .github/workflows/android.yaml [hunk 1]
```diff
@@ -46,3 +50,55 @@ jobs:
     with:
       platform: android-arm
       nightly: ${{ github.event.inputs.nightly }}
+      keep_artifacts: cobalt.apk
+
+  final_android_job:
```
Comment (oxve) [inline .github/workflows/android.yaml:55]: As mentioned in a comment below I think this needs an `if: ${{ github.event.release.name }}` or something along those lines.

File: .github/workflows/android.yaml [hunk 2]
```diff
@@ -46,3 +50,55 @@ jobs:
...
+          name: Android APKs
+          path: ./*
+          retention-days: 90
+          compression-level: 0 # We expect kept artifacts to be already compressed
+          if-no-files-found: error
+      - name: 'Delete intermediates'
+        uses: geekyeggo/delete-artifact@9d15d164b1dcd538ff1b1a2984bc2c0240986c3b # v4.0.0
+        with:
+          token: ${{ github.token }}
+          name: android-arm-qa
```
Comment (oxve) [inline .github/workflows/android.yaml:103]: Why are you uploading and then deleting `android-arm-qa`?

File: .github/workflows/main.yaml [hunk 3]
```diff
@@ -238,6 +243,15 @@ jobs:
         uses: ./.github/actions/gn
       - name: Build Cobalt
         uses: ./.github/actions/build
+      - name: 'Upload Artifact'
+        uses: actions/upload-artifact@v4
+        if: inputs.keep_artifacts
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply .github/workflows/main.yaml:248]: It's intended to run on every PR and postsubmit, not just releases.

The workflow is shared code and so i have to have a conditional in here.

File: .github/workflows/main.yaml [hunk 4]
```diff
@@ -29,6 +29,11 @@ on:
         required: false
         type: boolean
         default: false
+      keep_artifacts:
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply .github/workflows/main.yaml:32]: It expands to a path expression, so in principle multiple yes. For Android we just need this one, but for Evergreen-AOSP we'll need multiple ( elf_loader.apk and something else )

File: .github/workflows/android.yaml (hunk 2 above)
Comment (kaidokert) [reply .github/workflows/android.yaml:103]: I was intending to delete all intermediates actually, forgot. Unsure if this action actually works properly with artifact v4 though

File: .github/workflows/android.yaml (hunk 1 above)
Comment (kaidokert) [reply .github/workflows/android.yaml:55]: The intent is to run this on every PR, not gated on releases. For releases, extra code needs to be added at a later stage to post the results in Releases API.

File: .github/workflows/android.yaml (hunk 2 above)
Comment (kaidokert) [reply .github/workflows/android.yaml:103]: I'm getting an `Resource not accessible by integration` error when this runs.

The doc says:
```
Support for actions/upload-artifact@v4 utilizes the GitHub REST API, and requires a permissive [GITHUB_TOKEN](https://docs.github.com/en/actions/security-guides/automatic-token-authentication#permissions-for-the-github_token), or a PAT with read and write access to actions.
``` 

I thought we do have a permissive token already

File: .github/workflows/android.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (kaidokert) [reply .github/workflows/android.yaml]: Renamed

**Review by isarkis — APPROVED** [review]

**Pull request merged**